Blueprint to Restore Postgres data into RDS database Instance (#1768)
kanisterio/kanister: examples/postgres-RDS/README.md

# Postgres-RDS

[PostgreSQL](https://www.postgresql.org/) is an object-relational database management system (ORDBMS) with an emphasis on extensibility and standards compliance.

## Introduction

This example is to demonstrate how Kanister can be used to provision AWS RDS Instance and import data into it from On-prem Postgres using Kanister functions

## Prerequisites

- Kubernetes 1.10+
- PV provisioner support in the underlying infrastructure
- Kanister controller version 0.84.0 installed in your cluster
- Kanctl CLI installed (https://docs.kanister.io/tooling.html#kanctl)

## Create RDS instance on AWS

> We need to have a Postgres RDS instance where we would restore the data.

Reviewer: ```suggestion
We need to have a Postgres RDS instance where we will restore the data.
```

> You can skip this step if you already have an RDS instance created

> RDS instance needs to be reachable from the outside world. So make sure that you have VPC with the security group having the rule to allow ingress traffic on 5432 TCP port.

Reviewer: ```suggestion
RDS instance needs to be reachable from the Kubernetes cluster performing the restore operation. So make sure that you have VPC with the security group having the rule to allow ingress traffic on 5432 TCP port.
```

You can create a security group and add rules to the default VPC using the following commands

**NOTE**

> This is highly insecure. Its good only for testing and shouldn't be use in a production environment.
> 

Reviewer: ```suggestion
This is highly insecure. Its good only for testing and should not be use in a production environment.
For production, make sure to restrict the source (CIDR or otherwise) to only the elements that need to access the DB.
```

```bash
aws ec2 create-security-group --group-name pgtest-sg --description "pgtest security group"
aws ec2 authorize-security-group-ingress --group-name pgtest-sg --protocol tcp --port 5432 --cidr 0.0.0.0/0
```

Now create an RDS instance with the PostgreSQL engine

```bash
aws rds create-db-instance \
    --publicly-accessible \
    --allocated-storage 20 --db-instance-class db.t3.micro \
    --db-instance-identifier test-postgresql-instance \
    --engine postgres \

>     --master-username master \

Reviewer: ```suggestion
    --master-username <choose an admin username> \
```

    --vpc-security-group-id sg-xxxxyyyyzzz \ # Sec group with TCP 5432 inbound rule

>     --master-user-password secret99

Reviewer: ```suggestion
    --master-user-password <choose a secure password>
```

aws rds wait db-instance-available --db-instance-identifier=test-postgresql-instance
```
## Create configmap

Create a configmap that contains information to connect to the RDS DB instance

```
apiVersion: v1
kind: ConfigMap
metadata:
  name: dbconfig
data:
  region: us-west-2 #Region in which instance is to be created
  instance_name: test-postgresql-instance  # name of the RDS Instance that would be provisioned
```

## Create secret

Create a secret that contains credentials to connect to the RDS DB instance

```
apiVersion: v1
kind: Secret
metadata:
  name: dbsecret
stringData:
  ########## AWS Key CREDS
  accessKeyId: 
  secretAccessKey:
  ########## Database instance creds

>   postgres_username: master

Reviewer: ```suggestion
  postgres_username: <admin username chosen above>
```

>   postgres_password: secret99 

Reviewer: ```suggestion
  postgres_password: <secure password chosen above>
```

```

**NOTE**

> The creds must have proper permissions to perform operation on RDS

Reviewer: ```suggestion
The AWS credentials must have proper permissions to perform operations on RDS
```

$ kubectl --namespace kanister get actionsets.cr.kanister.io
NAME           AGE
backup-glptq   38s

# View the status of the actionset
$ kubectl --namespace kanister describe actionset backup-glptq
```

### Restore the Application to RDS Database instance

To restore the data into RDS Postgres instance, you should use the backup that you created before. An easy way to do this is to leverage `kanctl`, a command-line tool that helps create ActionSets that depend on other ActionSets. Apart from this we also need to pass RDS instance details to the blueprint, the secret and configmap resource containing these details can be passed to kanctl:

```bash
$ kanctl --namespace kanister create actionset --action restore --from backup-glptq --config-maps dbconfig=postgres-test/dbconfig --secrets dbsecret=postgres-test/dbsecret
actionset restore-backup-glptq-6jzt4 created

## Check status
$ kubectl --namespace kanister describe actionset restore-backup-glptq-6jzt4
```

Once the ActionSet status is set to "complete", you can see that the data has been successfully restored to the RDS PostgreSQL database instance

To verify, Connect to the RDS PostgreSQL database instance using `psql`

> 

Reviewer: ```suggestion
export PGHOST=$(aws rds describe-db-instances \
              --region us-west-2 \
              --db-instance-identifier test-postgresql-instance \
              --query "DBInstances[0].Endpoint.Address" \
              --output text)
```

> PGPASSWORD="secret99" psql --host test-postgresql-instance.cjbctojw4ahh.us-west-2.rds.amazonaws.com -U master -d postgres -p 5432

Reviewer: ```suggestion
PGPASSWORD="<secure password chosen above>" psql --host ${PGHOST} -U master -d postgres -p 5432
```

Add Kanister Helm repository and install Kanister operator
```bash
$ helm repo add kanister https://charts.kanister.io
$ helm install kanister --namespace kanister --create-namespace kanister/kanister-operator
```

## Integrating with Kanister

If you have deployed a PostgreSQL application with a name other than `my-release` and a namespace other than `postgres-test`, you need to modify the commands used below to use the correct name and namespace

### Create Profile

Create Profile CR if not created already

```bash
$ kanctl create profile s3compliant --access-key <aws-access-key-id> \
	--secret-key <aws-secret-key> \
	--bucket <s3-bucket-name> --region <region-name> \
	--namespace postgres-test
```

**NOTE:**

The command will configure a location where artifacts resulting from Kanister
data operations such as backup should go. This is stored as a `profiles.cr.kanister.io`
*CustomResource (CR)* which is then referenced in Kanister ActionSets. Every ActionSet
requires a Profile reference to complete the action. This CR (`profiles.cr.kanister.io`)
can be shared between Kanister-enabled application instances.

## Create Docker Image to be used in the Blueprint

> Create the docker image which include awscli, postgresclient and kando (https://docs.kanister.io/tooling.html#kando) . We would provide the image to the blueprint.

Reviewer 2: What is `postgresclient`? Do you mean postgres CLI?

Author: yeah, I meant `psql`

Reviewer 2: let's mention psql CLI instead of `postgresclient`.

Author: I have already modified it to `psql` there.

> Create the docker image which include awscli, postgresclient and kando (https://docs.kanister.io/tooling.html#kando) . We would provide the image to the blueprint.
> 

Reviewer 2: ```suggestion
Create the docker image which includes awscli, postgresclient and kando (https://docs.kanister.io/tooling.html#kando) . This image will be provided in the Kanister blueprint that we are going to create later.
```

----+------+-----+----------------------------------------------------+--------+----------------------------
 10 | Paul |  32 | California                                         |  20000 | 2019-09-16 14:39:36.316065
(1 row)

## Add few more entries
test=# INSERT INTO COMPANY (ID,NAME,AGE,ADDRESS,SALARY,CREATED_AT) VALUES (20, 'Omkar', 32, 'California', 20000.00, now());
INSERT 0 1
test=# INSERT INTO COMPANY (ID,NAME,AGE,ADDRESS,SALARY,CREATED_AT) VALUES (30, 'Prasad', 32, 'California', 20000.00, now());
INSERT 0 1

test=# select * from company;
 id | name  | age |                      address                       | salary |         created_at
----+-------+-----+----------------------------------------------------+--------+----------------------------
 10 | Paul  |  32 | California                                         |  20000 | 2019-09-16 14:39:36.316065
 20 | Omkar |  32 | California                                         |  20000 | 2019-09-16 14:40:52.952459
 30 | Omkar |  32 | California                                         |  20000 | 2019-09-16 14:41:06.433487
```

## Protect the Application

> You can now take a backup of the PostgresDB data using an ActionSet defining backup for this application. Create an ActionSet in the same namespace as the controller.
> 

Reviewer 2: ```suggestion
You can now take a backup of the PostgresDB data by creating backup ActionSet for this application. Create an ActionSet in the same namespace as the controller.

```

$ kubectl --namespace kanister get actionsets.cr.kanister.io
NAME           AGE
backup-glptq   38s

# View the status of the actionset
$ kubectl --namespace kanister describe actionset backup-glptq
```

### Restore the Application to RDS Database instance

To restore the data into RDS Postgres instance, you should use the backup that you created before. An easy way to do this is to leverage `kanctl`, a command-line tool that helps create ActionSets that depend on other ActionSets. Apart from this we also need to pass RDS instance details to the blueprint, the secret and configmap resource containing these details can be passed to kanctl:

```bash
$ kanctl --namespace kanister create actionset --action restore --from backup-glptq --config-maps dbconfig=postgres-test/dbconfig --secrets dbsecret=postgres-test/dbsecret
actionset restore-backup-glptq-6jzt4 created

## Check status
$ kubectl --namespace kanister describe actionset restore-backup-glptq-6jzt4
```

Once the ActionSet status is set to "complete", you can see that the data has been successfully restored to the RDS PostgreSQL database instance

To verify, Connect to the RDS PostgreSQL database instance 

```bash

> $ kubectl run my-release-postgresql-client --rm --tty -i --restart='Never' --namespace postgres-test --image docker.io/bitnami/postgresql:15.1.0-debian-11-r31 --env="PGPASSWORD=<db-password>" --command -- psql --host <instance-name> -U <master-username> -d template1 -p 5432                   

Reviewer 2: let's mention what should be the values of `instance-name`, and other placeholders.

Author: Do you mean using dummy values instead of placeholder? I modified these because of initial review comments I got [https://github.com/kanisterio/kanister/pull/1768#discussion_r1045580029](https://github.com/kanisterio/kanister/pull/1768#discussion_r1045580029) to replace values with placeholder.

Reviewer 2: > Do you mean using dummy values instead of placeholder?

What I meant is, either before or after the `kubectl` command, we should mention the `instance-name`, `db-password` and `master-username` are the details of the RDS instance that we have already created as part of this step (link to the step where we create the RDS instance).

Author: I have addressed that. PTAL
